### DISCUSSION TITLE
Options to produce a Descriptors.FileDescriptor out of a proto files with dependencies

### DISCUSSION THREAD

**Discussion by franz1981** [Apicurio/apicurio-registry#3819] [Show and tell]
Hi @carlesarnal !

This question/discussion is related https://github.com/Hyperfoil/Hyperfoil/issues/281, which requires to:
- parse a given proto file scanning its services information (including message types)
- allow to specify dependent proto files
- encode/decode dynamic messages based on the descriptions obtained from the proto(s) parsed

Our reference of user experience is analogous to https://ghz.sh/docs/example_config where the import paths are indeed a way to specify the additional proto files required to parse the main one ie `"proto"` in the mentioned documentation link.

By searching and modifying the `apicurio-registry-protobuf-schema-utilities` module, I have found 2 (not completed yet) ways to achieve what i need:
1. 2cf62b2565cbe6e28e1dcdb0d160892015781e43 which provide a new method
`Descriptors.FileDescriptor toFileDescriptor(protoFileName, protoFileElement, dependencies)` 
which requires specifying all the dependencies upfront (including well known types), although obtained without caring of parsing order (via `ProtoParser.Companion.parse`), **but** is not correctly populating the services in the parsed file descriptor (see [`testVertxGreeting`](https://github.com/Apicurio/apicurio-registry/commit/2cf62b2565cbe6e28e1dcdb0d160892015781e43#diff-c6c8173a1ad4ea73991e6e32b8943276398015c11cc8b34a79c383efe77f80f1R197) failure). The nice thing is that [`DynamicSchema::init`](https://github.com/franz1981/apicurio-registry/blob/2cf62b2565cbe6e28e1dcdb0d160892015781e43/utils/protobuf-schema-utilities/src/main/java/io/apicurio/registry/utils/protobuf/schema/DynamicSchema.java#L248) automatically resolve the provided dependencies.
2. 12166465866b6c3047ad5abb16f93d4e2ed425ef which is "fixing" an existing method ie [`loadSchema(packageName, fileName, schemaDefinition, deps)`](https://github.com/Apicurio/apicurio-registry/commit/12166465866b6c3047ad5abb16f93d4e2ed425ef#diff-a851fe83a30464e3940e5bacfd064fdcb3da172ce60891f7dd24895147187ad6R161) in order to allow `deps` maps to be placed in different folders than the main proto file loaded, see [`testProtoImportDifferentPackagesViaProtoFileToFileDescriptor`](https://github.com/Apicurio/apicurio-registry/commit/12166465866b6c3047ad5abb16f93d4e2ed425ef#diff-c6c8173a1ad4ea73991e6e32b8943276398015c11cc8b34a79c383efe77f80f1R385). This method seems to work fine with services definitions, but requires a full-fat map of dependencies, each in the `Descriptors.FileDescriptor` form, which means with their dependencies fully resolved too, if any (assuming I would use the same util method of the "main" proto). 
If compared to `1.`  this approach requires to implement my own dependencies resolution mechanism and ensure all the required `Descriptors.FileDescriptor` to be parsed upfront in the right order.

My questions are:
- do we have already a method that does what I need, elsewhere?
- if not, which of the 2 approaches is the better/more correct/idiomatic one?

**Comment by carlesarnal** [comment]
Hey there!

> Hi @carlesarnal !
> 
> This question/discussion is related [Hyperfoil/Hyperfoil#281](https://github.com/Hyperfoil/Hyperfoil/issues/281), which requires to:
> 
> * parse a given proto file scanning its services information (including message types)
> * allow to specify dependent proto files
> * encode/decode dynamic messages based on the descriptions obtained from the proto(s) parsed

Though one. 

> 
> Our reference of user experience is analogous to https://ghz.sh/docs/example_config where the import paths are indeed a way to specify the additional proto files required to parse the main one ie `"proto"` in the mentioned documentation link.
> 
> By searching and modifying the `apicurio-registry-protobuf-schema-utilities` module, I have found 2 (not completed yet) ways to achieve what i need:
> 
> 1. [2cf62b2](https://github.com/Apicurio/apicurio-registry/commit/2cf62b2565cbe6e28e1dcdb0d160892015781e43) which provide a new method
>    `Descriptors.FileDescriptor toFileDescriptor(protoFileName, protoFileElement, dependencies)`
>    which requires specifying all the dependencies upfront (including well known types), although obtained without caring of parsing order (via `ProtoParser.Companion.parse`), **but** is not correctly populating the services in the parsed file descriptor (see [`testVertxGreeting`](https://github.com/Apicurio/apicurio-registry/commit/2cf62b2565cbe6e28e1dcdb0d160892015781e43#diff-c6c8173a1ad4ea73991e6e32b8943276398015c11cc8b34a79c383efe77f80f1R197) failure). The nice thing is that [`DynamicSchema::init`](https://github.com/franz1981/apicurio-registry/blob/2cf62b2565cbe6e28e1dcdb0d160892015781e43/utils/protobuf-schema-utilities/src/main/java/io/apicurio/registry/utils/protobuf/schema/DynamicSchema.java#L248) automatically resolve the provided dependencies.
> 2. [1216646](https://github.com/Apicurio/apicurio-registry/commit/12166465866b6c3047ad5abb16f93d4e2ed425ef) which is "fixing" an existing method ie [`loadSchema(packageName, fileName, schemaDefinition, deps)`](https://github.com/Apicurio/apicurio-registry/commit/12166465866b6c3047ad5abb16f93d4e2ed425ef#diff-a851fe83a30464e3940e5bacfd064fdcb3da172ce60891f7dd24895147187ad6R161) in order to allow `deps` maps to be placed in different folders than the main proto file loaded, see [`testProtoImportDifferentPackagesViaProtoFileToFileDescriptor`](https://github.com/Apicurio/apicurio-registry/commit/12166465866b6c3047ad5abb16f93d4e2ed425ef#diff-c6c8173a1ad4ea73991e6e32b8943276398015c11cc8b34a79c383efe77f80f1R385). This method seems to work fine with services definitions, but requires a full-fat map of dependencies, each in the `Descriptors.FileDescriptor` form, which means with their dependencies fully resolved too, if any (assuming I would use the same util method of the "main" proto).
>    If compared to `1.`  this approach requires to implement my own dependencies resolution mechanism and ensure all the required `Descriptors.FileDescriptor` to be parsed upfront in the right order.
> 
> My questions are:
> 
> * do we have already a method that does what I need, elsewhere?

Kind of, we do have a maven-plugin that does the dependency resolution, and the code can probably be extracted from there to the protobuf utilities library, for reference https://github.com/Apicurio/apicurio-registry/blob/main/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/ProtobufDirectoryParser.java#L37.

> * if not, which of the 2 approaches is the better/more correct/idiomatic one?

I believe 2 is the way to go here (that's why we have the resolution on the maven-plugin :)). In the end, if you have a proto definition with a dependency and you want to produce messages/requests in the end you're going to need the dependency definition as well. 

That said, we recently discovered that we might have a [bug](https://github.com/Apicurio/apicurio-registry/issues/3762) resolving proto files with different packages, at least in runtime server (we're always talking about client-side operation from our perspective on this discussion), I'll keep you posted.

**Reply by franz1981** [reply]
Thanks for the response! I'll than stick with the second approach (and look at the plugin code) and prepare a PR to discuss the API/impl I'm using! many thanks!
